File: libs/event-bus-control/src/connectors.rs
use serde_json::{Value, json};

use crate::schema_registry::{self, SchemaType};

#[derive(Debug, Clone)]
pub struct EventStreamTopic {
    pub selector: String,
    pub display_name: String,
    pub sample_messages: Vec<Value>,
    pub partitions: i64,
    pub metadata: Value,
}

pub fn validate_topic_connector_config(
    config: &Value,
    connector_label: &str,
) -> Result<(), String> {
    if bootstrap_servers(config).is_none() {
        return Err(format!(
            "{connector_label} requires 'bootstrap_servers' or 'brokers'"
        ));
    }

    let topics = parse_topic_entries(config, connector_label)?;
    if topics.is_empty() {
        return Err(format!(
            "{connector_label} requires at least one topic in 'topics'"
        ));
    }

    // If a topic ships an inline schema (Avro / JSON Schema / Protobuf
    // descriptor set), validate the configured `sample_messages` against
    // it. This catches schema/payload drift at registration time instead
    // of at first read.
    for topic in &topics {
        validate_topic_samples(&topic.metadata, &topic.sample_messages, connector_label)?;
    }

    Ok(())
}

pub fn parse_topic_entries(
    config: &Value,
    connector_label: &str,
) -> Result<Vec<EventStreamTopic>, String> {
    let topics = config
        .get("topics")
        .and_then(Value::as_array)
        .ok_or_else(|| format!("{connector_label} requires 'topics' to be an array"))?;

    let mut parsed = Vec::with_capacity(topics.len());
    for (index, topic) in topics.iter().enumerate() {
        if let Some(selector) = topic
            .as_str()
            .map(str::trim)
            .filter(|value| !value.is_empty())
        {
            parsed.push(EventStreamTopic {
                selector: selector.to_string(),
                display_name: selector.to_string(),
                sample_messages: Vec::new(),
                partitions: 1,
                metadata: json!({ "topic": selector }),
            });
            continue;
        }

        let selector = topic
            .get("selector")
            .or_else(|| topic.get("topic"))
            .or_else(|| topic.get("name"))
            .and_then(Value::as_str)
            .filter(|value| !value.trim().is_empty())
            .ok_or_else(|| {
                format!("{connector_label} topics[{index}] requires 'selector', 'topic' or 'name'")
            })?
            .to_string();
        let display_name = topic
            .get("display_name")
            .or_else(|| topic.get("name"))
            .and_then(Value::as_str)
            .unwrap_or(&selector)
            .to_string();
        let sample_messages = topic
            .get("sample_messages")
            .or_else(|| topic.get("preview_rows"))
            .and_then(Value::as_array)
            .cloned()
            .unwrap_or_default();
        let partitions = topic
            .get("partitions")
            .and_then(Value::as_i64)
            .unwrap_or(1)
            .max(1);
        let mut metadata = topic.clone();
        if let Some(object) = metadata.as_object_mut() {
            object.remove("sample_messages");
            object.remove("preview_rows");
        }

        parsed.push(EventStreamTopic {
            selector,
            display_name,
            sample_messages,
            partitions,
            metadata,
        });
    }

    Ok(parsed)
}

pub fn find_topic_entry(
    config: &Value,
    selector: &str,
    connector_label: &str,
) -> Result<EventStreamTopic, String> {
    parse_topic_entries(config, connector_label)?
        .into_iter()
        .find(|topic| topic.selector == selector)
        .ok_or_else(|| format!("{connector_label} topic '{selector}' is not configured"))
}

/// Validate a topic's `sample_messages` against an inline schema declared
/// in the topic config. The expected shape is:
///
/// ```json
/// {
///   "selector": "orders",
///   "schema": { "type": "avro" | "protobuf" | "json", "text": "..." },
///   "sample_messages": [ { ... } ]
/// }
/// ```
///
/// `schema_subject` (a reference to a registered subject in the
/// cdc-metadata-service Schema Registry) is recognised but only used by
/// the connector for traceability — actual validation always runs against
/// the inline `schema.text` when present. This keeps the connector usable
/// in environments where the registry is offline (catalog_backed mode).
///
/// Returns `Ok(())` if no schema is configured (validation is opt-in).
pub fn validate_topic_samples(
    topic_metadata: &Value,
    sample_messages: &[Value],
    connector_label: &str,
) -> Result<(), String> {
    let Some(schema) = topic_metadata.get("schema").filter(|value| !value.is_null()) else {
        return Ok(());
    };
    let schema_type_str = schema
        .get("type")
        .and_then(Value::as_str)
        .ok_or_else(|| format!("{connector_label} schema requires 'type'"))?;
    let schema_type: SchemaType = schema_type_str
        .parse()
        .map_err(|error: schema_registry::SchemaError| {
            format!("{connector_label} {error}")
        })?;
    let schema_text = schema
        .get("text")
        .and_then(Value::as_str)
        .ok_or_else(|| format!("{connector_label} schema requires 'text'"))?;
    for (index, sample) in sample_messages.iter().enumerate() {
        schema_registry::validate_payload(schema_type, schema_text, sample).map_err(|error| {
            format!(
                "{connector_label} sample_messages[{index}] does not match schema: {error}"
            )
        })?;
    }
    Ok(())
}

pub fn bootstrap_servers(config: &Value) -> Option<&str> {
    string_field(config, "bootstrap_servers").or_else(|| string_field(config, "brokers"))
}

pub fn sanitize_file_stem(selector: &str, fallback: &str) -> String {
    let stem = selector
        .chars()
        .map(|ch| if ch.is_ascii_alphanumeric() { ch } else { '_' })
        .collect::<String>()
        .trim_matches('_')
        .to_string();
    if stem.is_empty() {
        fallback.to_string()
    } else {
        stem.chars().take(64).collect()
    }
}

fn string_field<'a>(config: &'a Value, field: &str) -> Option<&'a str> {
    config
        .get(field)
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::{
        bootstrap_servers, find_topic_entry, parse_topic_entries, sanitize_file_stem,
        validate_topic_connector_config,
    };

    #[test]
    fn parses_string_and_object_topics() {
        let topics = parse_topic_entries(
            &json!({
                "topics": [
                    "orders",
                    {
                        "selector": "payments",
                        "display_name": "Payments",
                        "partitions": 6,
                        "sample_messages": [{ "payment_id": "pay-1" }]
                    }
                ]
            }),
            "kafka connector",
        )
        .expect("topics should parse");

        assert_eq!(topics.len(), 2);
        assert_eq!(topics[0].selector, "orders");
        assert_eq!(topics[1].display_name, "Payments");
        assert_eq!(topics[1].partitions, 6);
    }

    #[test]
    fn validates_required_bootstrap_servers() {
        let error =
            validate_topic_connector_config(&json!({ "topics": ["orders"] }), "kafka connector")
                .expect_err("validation should fail");
        assert!(error.contains("bootstrap_servers"));
    }

    #[test]
    fn finds_configured_topic() {
        let topic = find_topic_entry(
            &json!({
                "bootstrap_servers": "broker-a:9092",
                "topics": [
                    {
                        "selector": "orders",
                        "sample_messages": [{ "order_id": "ord-1" }]
                    }
                ]
            }),
            "orders",
            "kafka connector",
        )
        .expect("topic should exist");

        assert_eq!(topic.selector, "orders");
        assert_eq!(topic.sample_messages, vec![json!({ "order_id": "ord-1" })]);
        assert_eq!(
            bootstrap_servers(&json!({ "brokers": "broker-a:9092" })),
            Some("broker-a:9092")
        );
    }

    #[test]
    fn sanitizes_file_stems() {
        assert_eq!(sanitize_file_stem("orders.v1", "fallback"), "orders_v1");
        assert_eq!(sanitize_file_stem("///", "fallback"), "fallback");
    }

    #[test]
    fn topic_with_inline_avro_schema_validates_samples() {
        let config = json!({
            "bootstrap_servers": "broker-a:9092",
            "topics": [{
                "selector": "orders",
                "schema": {
                    "type": "avro",
                    "text": r#"{"type":"record","name":"Order","fields":[{"name":"order_id","type":"string"}]}"#
                },
                "sample_messages": [{ "order_id": "ord-1" }]
            }]
        });
        validate_topic_connector_config(&config, "kafka connector").expect("samples match schema");
    }

    #[test]
    fn topic_with_inline_schema_rejects_invalid_sample() {
        let config = json!({
            "bootstrap_servers": "broker-a:9092",
            "topics": [{
                "selector": "orders",
                "schema": {
                    "type": "json",
                    "text": r#"{"type":"object","required":["order_id"],"properties":{"order_id":{"type":"string"}}}"#
                },
                "sample_messages": [{ "order_id": "ord-1" }, { "wrong_field": 1 }]
            }]
        });
        let error = validate_topic_connector_config(&config, "kafka connector")
            .expect_err("second sample is missing required field");
        assert!(error.contains("sample_messages[1]"));
    }
}

File: openfoundry-go/libs/event-bus-control/connectors.go
package controlbus

// connectors.go ports libs/event-bus-control/src/connectors.rs.
//
// Connector-config helpers shared by data-connection plane modules.
// Pure logic — no DB, no HTTP. Inputs are JSON-decoded config blocks
// from the catalog and the helpers either parse them into typed
// structs (EventStreamTopic) or validate that the operator's
// configuration is internally consistent (sample messages match
// inline schemas, bootstrap servers are present, etc.).

import (
	"encoding/json"
	"fmt"
	"strings"
)

// EventStreamTopic is one parsed entry from a connector's `topics`
// array. Mirrors the Rust EventStreamTopic struct.
type EventStreamTopic struct {
	Selector       string
	DisplayName    string
	SampleMessages []json.RawMessage
	Partitions     int64
	Metadata       json.RawMessage
}

// ValidateTopicConnectorConfig validates a Kafka-style connector
// config: requires bootstrap_servers (or brokers) plus at least one
// topic, then runs ValidateTopicSamples on each topic that ships an
// inline schema.
//
// connectorLabel is used in error messages so callers can identify
// which connector failed when many are batch-validated.
func ValidateTopicConnectorConfig(config json.RawMessage, connectorLabel string) error {
	bootstrap, _ := BootstrapServers(config)
	if bootstrap == "" {
		return fmt.Errorf("%s requires 'bootstrap_servers' or 'brokers'", connectorLabel)
	}
	topics, err := ParseTopicEntries(config, connectorLabel)
	if err != nil {
		return err
	}
	if len(topics) == 0 {
		return fmt.Errorf("%s requires at least one topic in 'topics'", connectorLabel)
	}
	for _, topic := range topics {
		if err := ValidateTopicSamples(topic.Metadata, topic.SampleMessages, connectorLabel); err != nil {
			return err
		}
	}
	return nil
}

// ParseTopicEntries decodes the connector's `topics` field into a
// slice of EventStreamTopic. Accepts both the short form (string)
// and the long form (object with selector/display_name/...).
func ParseTopicEntries(config json.RawMessage, connectorLabel string) ([]EventStreamTopic, error) {
	var raw struct {
		Topics []json.RawMessage `json:"topics"`
	}
	if err := json.Unmarshal(config, &raw); err != nil {
		return nil, fmt.Errorf("%s requires 'topics' to be an array", connectorLabel)
	}
	if raw.Topics == nil {
		return nil, fmt.Errorf("%s requires 'topics' to be an array", connectorLabel)
	}
	out := make([]EventStreamTopic, 0, len(raw.Topics))
	for index, topicBytes := range raw.Topics {
		if entry, ok := tryParseStringTopic(topicBytes); ok {
			out = append(out, entry)
			continue
		}
		entry, err := parseObjectTopic(topicBytes, index, connectorLabel)
		if err != nil {
			return nil, err
		}
		out = append(out, entry)
	}
	return out, nil
}

// FindTopicEntry returns the topic with `selector` from `config`.
func FindTopicEntry(config json.RawMessage, selector, connectorLabel string) (EventStreamTopic, error) {
	topics, err := ParseTopicEntries(config, connectorLabel)
	if err != nil {
		return EventStreamTopic{}, err
	}
	for _, t := range topics {
		if t.Selector == selector {
			return t, nil
		}
	}
	return EventStreamTopic{}, fmt.Errorf("%s topic '%s' is not configured", connectorLabel, selector)
}

// ValidateTopicSamples validates a topic's sample_messages against an
// inline schema declared in the topic config. Expected metadata shape:
//
//	{
//	  "selector": "orders",
//	  "schema":   { "type": "avro" | "protobuf" | "json", "text": "..." },
//	  "sample_messages": [ { ... } ]
//	}
//
// schema_subject (a reference to a registered subject in the
// ingestion-replication-service Schema Registry) is recognised but
// only used by the connector for traceability — actual validation
// always runs against the inline schema.text when present.
//
// Returns nil if no schema is configured (validation is opt-in).
func ValidateTopicSamples(topicMetadata json.RawMessage, samples []json.RawMessage, connectorLabel string) error {
	var meta map[string]json.RawMessage
	if err := json.Unmarshal(topicMetadata, &meta); err != nil {
		// Metadata not an object — nothing to validate against.
		return nil
	}
	schemaRaw, ok := meta["schema"]
	if !ok || isNullJSON(schemaRaw) {
		return nil
	}
	var schema struct {
		Type string `json:"type"`
		Text string `json:"text"`
	}
	if err := json.Unmarshal(schemaRaw, &schema); err != nil {
		return fmt.Errorf("%s schema: %s", connectorLabel, err.Error())
	}
	if schema.Type == "" {
		return fmt.Errorf("%s schema requires 'type'", connectorLabel)
	}
	schemaType, err := ParseSchemaType(schema.Type)
	if err != nil {
		return fmt.Errorf("%s %s", connectorLabel, err.Error())
	}
	if schema.Text == "" {
		return fmt.Errorf("%s schema requires 'text'", connectorLabel)
	}
	for index, sample := range samples {
		if err := ValidatePayload(schemaType, schema.Text, sample); err != nil {
			return fmt.Errorf("%s sample_messages[%d] does not match schema: %s",
				connectorLabel, index, err.Error())
		}
	}
	return nil
}

// BootstrapServers returns the connector's broker bootstrap string.
// Accepts both `bootstrap_servers` and `brokers` keys (Confluent vs
// standard Kafka spelling).
func BootstrapServers(config json.RawMessage) (string, bool) {
	if v, ok := stringField(config, "bootstrap_servers"); ok {
		return v, true
	}
	return stringField(config, "brokers")
}

// SanitizeFileStem normalises a topic selector into a filesystem-safe
// filename stem (alphanumerics only, leading/trailing underscores
// trimmed, max 64 chars). Falls back to `fallback` when the result
// is empty.
func SanitizeFileStem(selector, fallback string) string {
	out := make([]byte, 0, len(selector))
	for _, r := range selector {
		if (r >= '0' && r <= '9') || (r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z') {
			out = append(out, byte(r))
		} else {
			out = append(out, '_')
		}
	}
	stem := strings.Trim(string(out), "_")
	if stem == "" {
		return fallback
	}
	if len(stem) > 64 {
		stem = stem[:64]
	}
	return stem
}

// ─── internals ─────────────────────────────────────────────────────────

func tryParseStringTopic(raw json.RawMessage) (EventStreamTopic, bool) {
	var s string
	if err := json.Unmarshal(raw, &s); err != nil {
		return EventStreamTopic{}, false
	}
	trimmed := strings.TrimSpace(s)
	if trimmed == "" {
		return EventStreamTopic{}, false
	}
	metadata, _ := json.Marshal(map[string]string{"topic": trimmed})
	return EventStreamTopic{
		Selector:       trimmed,
		DisplayName:    trimmed,
		SampleMessages: []json.RawMessage{},
		Partitions:     1,
		Metadata:       metadata,
	}, true
}

func parseObjectTopic(raw json.RawMessage, index int, connectorLabel string) (EventStreamTopic, error) {
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(raw, &obj); err != nil {
		return EventStreamTopic{}, fmt.Errorf(
			"%s topics[%d] requires 'selector', 'topic' or 'name'", connectorLabel, index)
	}

	selector := pickStringField(obj, "selector", "topic", "name")
	selector = strings.TrimSpace(selector)
	if selector == "" {
		return EventStreamTopic{}, fmt.Errorf(
			"%s topics[%d] requires 'selector', 'topic' or 'name'", connectorLabel, index)
	}

	displayName := pickStringField(obj, "display_name", "name")
	if displayName == "" {
		displayName = selector
	}

	var samples []json.RawMessage
	if rawSamples, ok := obj["sample_messages"]; ok && !isNullJSON(rawSamples) {
		_ = json.Unmarshal(rawSamples, &samples)
	} else if rawSamples, ok := obj["preview_rows"]; ok && !isNullJSON(rawSamples) {
		_ = json.Unmarshal(rawSamples, &samples)
	}
	if samples == nil {
		samples = []json.RawMessage{}
	}

	partitions := int64(1)
	if rawParts, ok := obj["partitions"]; ok {
		var n int64
		if err := json.Unmarshal(rawParts, &n); err == nil && n > 1 {
			partitions = n
		}
	}

	// Build metadata = clone(topicObject) minus sample_messages /
	// preview_rows. Mirrors the Rust impl exactly.
	metaCopy := make(map[string]json.RawMessage, len(obj))
	for k, v := range obj {
		if k == "sample_messages" || k == "preview_rows" {
			continue
		}
		metaCopy[k] = v
	}
	metadata, err := json.Marshal(metaCopy)
	if err != nil {
		metadata = json.RawMessage(`{}`)
	}

	return EventStreamTopic{
		Selector:       selector,
		DisplayName:    displayName,
		SampleMessages: samples,
		Partitions:     partitions,
		Metadata:       metadata,
	}, nil
}

// pickStringField returns the first non-empty string field amongst
// `keys`. Empty string when none matched.
func pickStringField(obj map[string]json.RawMessage, keys ...string) string {
	for _, k := range keys {
		raw, ok := obj[k]
		if !ok {
			continue
		}
		var s string
		if err := json.Unmarshal(raw, &s); err == nil && s != "" {
			return s
		}
	}
	return ""
}

func stringField(config json.RawMessage, field string) (string, bool) {
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(config, &obj); err != nil {
		return "", false
	}
	raw, ok := obj[field]
	if !ok {
		return "", false
	}
	var s string
	if err := json.Unmarshal(raw, &s); err != nil {
		return "", false
	}
	if strings.TrimSpace(s) == "" {
		return "", false
	}
	return s, true
}

func isNullJSON(raw json.RawMessage) bool {
	return len(raw) == 0 || string(raw) == "null"
}
